account-settings: Disable deactivate account if user is the only owner.

laurynmm opened the pull request:
Disables the deactivate account button in the user's account and privacy settings tab if they are the only organization owner. Adds a tooltip with the 'fa fa-info-circle' icon when the button is deactivated.

See [relevant CZO conversation](https://chat.zulip.org/#narrow/stream/6-frontend/topic/deactivate.20user.20button.2Fmodal/near/1474195).

**Notes**:
- Loops over the active users in the organization to get the count of org owners, which might have performance issues in organizations with a large number of active users. Not sure what a reasonable `populate_db` with extra users for testing this in the dev environment might be.
- The variable name `only_organization_owner` made sense to me in the template with the if/unless statements, but there might be a clearer naming option I didn't consider.

---

**Screenshots and screen captures:**

<details>
<summary>Deactivate account button - tooltip if disabled</summary>

![Screenshot from 2022-12-08 16-42-07](https://user-images.githubusercontent.com/63245456/206495277-e2e90c98-7807-4e0d-999b-667d68c6f1fc.png)
</details>
<details>
<summary>Deactivate account button - disabled</summary>

![Screenshot from 2022-12-08 16-42-13](https://user-images.githubusercontent.com/63245456/206495285-7ff43fee-c685-4b3c-bcda-7d7d470c6334.png)
</details>
<details>
<summary>Compare Iago (left) who is admin to Desdemona (right) who is the only organization admin</summary>

![Screenshot from 2022-12-08 16-42-55](https://user-images.githubusercontent.com/63245456/206495287-e09afe1c-22db-4969-9990-aa7a064e978f.png)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on static/js/people.js:
```diff
@@ -884,6 +884,16 @@ export function get_active_human_count() {
+            count += 1;
+        }
+    }
+    return count;
```
Can we just have this function return whether it is the last owner or not? We currently don't need count of owners in the realm.

laurynmm replied:
That makes sense. I've updated for this to `is_current_user_only_owner`, which returns a boolean. I had a question about that new function [here](https://github.com/zulip/zulip/pull/23797/files/e8181d5404ba9170c584b5c82bfa17c05179d41a#r1046179965).

laurynmm commented on static/js/people.js:
```diff
@@ -825,6 +825,23 @@ export function is_active_user_for_popover(user_id) {
+            return false;
+        }
+    }
+    return true;
```
If `active_owners` is not equal to 1 here, then there is an error in our data, because the current user has `page_params.is_owner = true`.

Would it make sense to add a blueslip warning or error here for that unlikely case?

timabbott replied:
It's a reasonable thing to do, certainly. But the most likely bug in such a situation would be a server bug, so I don't think it's important to do so.

laurynmm commented:
*Note that the failing puppeteer test is not related to these changes.*

**Updates**:
- Updated the logic / check for whether the current user is the only organization owner based on review feedback.
- Added live update for when user's roles are updated.
- Moved the disabled tooltip text to be above the button text, [this feedback](https://github.com/zulip/zulip/pull/23797#issuecomment-1343200453). But ideally will update to be like #23810 when that gets reviewed/merged.

<details>
<summary>Updated screenshot of disabled deactivate account button with text</summary>

![Screenshot from 2022-12-12 18-45-37](https://user-images.githubusercontent.com/63245456/207132476-5bb9637c-1801-49b6-8b1a-62f796d2ed99.png)
</details>

laurynmm commented on static/js/server_events_dispatch.js:
```diff
@@ -429,15 +429,18 @@ export function dispatch_normal_event(event) {
                     break;
                 case "update":
                     user_events.update_person(event.person);
+                    settings_account.maybe_update_deactivate_account_button();
```
We want to check for this update for all three cases (add, remove and update), but necessarily for the error case.

sahil839 replied:
There are a couple of places where we call the common function for all the cases whereas for one case we call it once (and thus is called in the error case as well). Not sure what we prefer, but I find this, i.e. calling the function for all cases,  more clearer.

laurynmm commented:
**Updates**:
- Moved the information about why the button is disabled to a tooltip when hovering on the disabled button. The tooltip appears to the right (see screenshots below), since it doesn't overlap anything there.
- Live updates the button state and tooltip if another organization owner is added/reactivated, deactivated or has their role changed/updated.
- I have two outstanding questions / comments noted above.

---

**Screenshots**:
<details>
<summary>Multiple organization owners</summary>

![Screenshot from 2022-12-20 19-31-47](https://user-images.githubusercontent.com/63245456/208741527-e8e9879b-0689-4ef2-adea-e9dd2adc5b04.png)
</details>
<details>
<summary>One organization owner - full screen</summary>

![Screenshot from 2022-12-20 19-42-32](https://user-images.githubusercontent.com/63245456/208742241-890ac3e3-6602-485b-bcda-eea1cf145160.png)
</details>
<details>
<summary>One organization owner - narrow screen</summary>

![Screenshot from 2022-12-20 19-42-11](https://user-images.githubusercontent.com/63245456/208742273-a65fad9b-0799-4842-8bba-35d9eb110cd9.png)
</details>

alya commented:
The screenshots look good to me!

@sahil839 do you have feedback on the questions above, or should @timabbott take a look next?

sahil839 commented:
Commented my thoughts on one of the questions above, not sure about the other one.

laurynmm commented:
@alya - I made one update on this that you might want to approve before having @timabbott review.

The tooltip to the right of the button does not work for mobile screens when I use the browser dev tools, so it now appears in the default location above the button, updated screenshots below.

I added screenshots of the tooltip to the right and under the button, so that you can see why I made the change and another option. When the tooltip is under the button, it covers the **Privacy** section title. When it's above the button, it covers the password field.

---

**Screenshots**:

### Current PR tooltip location

<details>
<summary>Full screen in web-app</summary>

![Screenshot from 2023-01-02 18-07-05](https://user-images.githubusercontent.com/63245456/210261670-b779db35-8975-4b48-86a7-9a71f33538e9.png)
</details>
<details>
<summary>Narrow screen in web-app</summary>

![Screenshot from 2023-01-02 18-07-30](https://user-images.githubusercontent.com/63245456/210261659-0f30004c-d23e-4a98-835a-a8020d0ef3b9.png)
</details>
<details>
<summary>mobile screen with dev tools</summary>

![Screenshot from 2023-01-02 18-07-55](https://user-images.githubusercontent.com/63245456/210261606-d3f79a12-dea0-477a-a5e9-b0e99ab74680.png)
</details>

### Alternative tooltip locations

<details>
<summary>mobile screen with tooltip to right of button - off screen</summary>

![Screenshot from 2023-01-02 18-12-31](https://user-images.githubusercontent.com/63245456/210261888-b248ddef-c676-4890-a03c-f05b7180a1f2.png)
</details>
<details>
<summary>mobile screen with tooltip under button</summary>

![Screenshot from 2023-01-02 18-14-50](https://user-images.githubusercontent.com/63245456/210262057-c43d329e-6542-4975-b713-5e494c1c402c.png)
</details>
<details>
<summary>Full screen with tooltip under button</summary>

![Screenshot from 2023-01-02 18-19-29](https://user-images.githubusercontent.com/63245456/210262490-7ab7e731-5501-4496-8fd3-6b173a04405a.png)

</details>

alya commented:
Hm, I don't think it's worth worrying a ton about -- I liked the right-side placement, but using the default is OK too.

However, note that there is no hover action on actual mobile devices, so it might not matter if the tooltip gets hidden at that screen size.

timabbott commented:
This looks good as an implementation. 

I think 95% of the time when you're the only owner of an organization, and you're trying to deactivate your account, you probably want to deactivate the organization and will find this message a bit unhelpful. We should extend the tooltip to have instructions, or perhaps better if that fits into a tooltpi, a link?
 
Maybe @alya can open an issue with a design for this use case -- I'm not sure what's best and it seems like a fine follow-up.

Merged, thanks @laurynmm!

alya commented:
I filed #24105 as a follow-up issue.
